File: 11-stacks-queues2/5-problem-applications/main.py
# Problem Applications - Python Example

def largest_rectangle_area(heights):
    stack = []
    max_area = 0
    heights.append(0)
    for i, h in enumerate(heights):
        while stack and heights[stack[-1]] > h:
            height = heights[stack.pop()]
            width = i if not stack else i - stack[-1] -1
            max_area = max(max_area, height * width)
        stack.append(i)
    return max_area

heights = [2,1,5,6,2,3]
print("Largest rectangle area:", largest_rectangle_area(heights))

File: 11-stacks-queues2/5-problem-applications/main.java
// Problem Applications - Java Example

import java.util.Stack;

public class ProblemApplications {

    public static int largestRectangleArea(int[] heights) {
        Stack<Integer> stack = new Stack<>();
        int maxArea = 0;
        int[] extendedHeights = new int[heights.length + 1];
        System.arraycopy(heights, 0, extendedHeights, 0, heights.length);

        for (int i = 0; i < extendedHeights.length; i++) {
            while (!stack.isEmpty() && extendedHeights[stack.peek()] > extendedHeights[i]) {
                int height = extendedHeights[stack.pop()];
                int width = stack.isEmpty() ? i : i - stack.peek() - 1;
                maxArea = Math.max(maxArea, height * width);
            }
            stack.push(i);
        }
        return maxArea;
    }

    public static void main(String[] args) {
        int[] heights = {2,1,5,6,2,3};
        System.out.println("Largest rectangle area: " + largestRectangleArea(heights));
    }
}
